### PR TITLE
Save recovery code pdf feature

### PR DESCRIPTION
**Pull request by cmonfortep**
Task/Issue URL: https://app.asana.com/0/488551667048375/1204316129840762/f 

### Description
Implements storing the recovery code as PDF and share action.

### Steps to test this PR

_Feature 1_
- [x] Remove `FakeDeviceSyncState` class
- [x] run the app
- [x] create a new sync account (stop on the "save recovery PDF? screen")
- [x] on the screen "save recovery PDF?"
- [x] click on Save recovery code
- [x] should prompt for permission if required
- [x] accept permission
- [x] should prompt share with the generated pdf (it will be stored in downloads as well)
- [x] share the pdf with another app to validate the pdf contains QR code and recovery code in B64
- [x] Go to sync main screen
- [x] click on Save recovery PDF
- [x] should prompt for permission if required
- [x] accept permission
- [x] should prompt share with the generated pdf (it will be stored in downloads as well)
- [x] share the pdf with another app to validate the pdf contains QR code and recovery code in B64

### UI changes
| Before  | After |
| ------ | ----- |
!(Upload before screenshot)|(Upload after screenshot)|

### CODE REVIEW

**Comment by cmonfortep** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #3063** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/3063" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/duckduckgo/Android/3063?utm_source=stack-comment).

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/RecoveryCodePDF.kt [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+import com.duckduckgo.mobile.android.ui.view.toPx
+import com.duckduckgo.sync.impl.databinding.ViewRecoveryCodeBinding
+import com.squareup.anvil.annotations.ContributesBinding
+import java.io.File
+import java.io.FileOutputStream
+import javax.inject.*
+import kotlinx.coroutines.withContext
+
+interface RecoveryCodePDF {
+    suspend fun generateAndStoreRecoveryCodePDF(
```
Comment (malmstein) [inline sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/RecoveryCodePDF.kt:36]: this doesn’t need to be `suspend`, adding a `@WorkerThread` annotation would be better. The `ViewModel` can/should deal with threading

**Review by malmstein — APPROVED** [review]
Nothing blocking but I’ve added a few suggestions from an action responsibility point of view

File: sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/ui/setup/SaveRecoveryCodeFragment.kt [hunk 2]
```diff
@@ -91,6 +113,14 @@ class SaveRecoveryCodeFragment : DuckDuckGoFragment(R.layout.fragment_recovery_c
         when (it) {
             Error -> requireActivity().finish()
             Finish -> requireActivity().finish()
+            is Command.StoreRecoveryCodePDF -> {
+                storagePermission.invokeOrRequestPermission {
+                    lifecycleScope.launch {
+                        val generateRecoveryCodePDF = recoveryCodePDF.generateAndStoreRecoveryCodePDF(requireContext(), it.recoveryCodeB64)
```
[the remark this answers is not in the document]
Comment (cmonfortep) [reply sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/ui/setup/SaveRecoveryCodeFragment.kt:123]: I need a view context to generate the pdf, that's the main reason I have this code here. I can move it to the ViewModel, and pass the view context from the UI. I agree is ViewModel responsibility, but without being able to inject the view context, I was not convinced.

File: sync/sync-impl/src/main/res/layout/view_recovery_code.xml [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+        <LinearLayout
+            android:layout_width="0dp"
+            android:layout_height="wrap_content"
+            android:orientation="vertical"
+            android:layout_weight="1"
+            android:layout_marginStart="@dimen/keyline_4"
+            android:layout_marginEnd="@dimen/keyline_4">
+            <com.duckduckgo.mobile.android.ui.view.text.DaxTextView
+                android:layout_width="36dp"
+                android:layout_height="36sp"
```
[the remark this answers is not in the document]
Comment (cmonfortep) [reply sync/sync-impl/src/main/res/layout/view_recovery_code.xml:221]: eagle eye

File: sync/sync-impl/src/main/res/drawable/circled_accent_background.xml [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  ~ Unless required by applicable law or agreed to in writing, software
+  ~ distributed under the License is distributed on an "AS IS" BASIS,
+  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+
+<shape xmlns:android="http://schemas.android.com/apk/res/android">
+    <solid android:color="?attr/daxColorAccentBlue" />
+    <corners android:radius="1000dp" /> <!-- Set the radius to a large value to make it look like a circle -->
```
[the remark this answers is not in the document]
Comment (cmonfortep) [reply sync/sync-impl/src/main/res/drawable/circled_accent_background.xml:20]: `oval` works 👍

**Pull request merged**